capsules/extra/src/distance.rs

```diff
@@ -0,0 +1,195 @@
+//! );
+//! virtual_alarm_hc_sr04.setup();
+//!
+//! let hc_sr04 = static_init!(
+//!     hc_sr04::HcSr04<VirtualMuxAlarm<'static, RPTimer>>,
+//!     hc_sr04::HcSr04::new(trig_pin, echo_pin, virtual_alarm_hc_sr04)
+//! );
+//! virtual_alarm_hc_sr04.set_alarm_client(hc_sr04);
+//!
+//! echo_pin.set_client(hc_sr04);
+//!
+//! let distance_sensor = static_init!(
+//!     distance::DistanceSensor<
+//!         'static,
+//!         hc_sr04::HcSr04<'static, VirtualMuxAlarm<'static, RPTimer>>,
+//!     >,
+//!     distance::DistanceSensor::new(
+//!         hc_sr04,
+//!         board_kernel.create_grant(distance::DRIVER_NUM, &memory_allocation_capability)
+//!     )
+//! );
+//! hc_sr04.set_client(distance_sensor);
+//! ```
+
+use core::cell::Cell;
+
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::hil;
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Distance as usize;
+
+#[derive(Default)]
+pub struct App {
+    subscribed: bool,
+}
+
+pub struct DistanceSensor<'a, T: hil::sensors::Distance<'a>> {
+    driver: &'a T,
+    apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    busy: Cell<bool>,
+}
+
+impl<'a, T: hil::sensors::Distance<'a>> DistanceSensor<'a, T> {
+    pub fn new(
+        driver: &'a T,
+        grant: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> DistanceSensor<'a, T> {
+        DistanceSensor {
+            driver,
+            apps: grant,
+            busy: Cell::new(false),
+        }
+    }
+
+    fn enqueue_command(&self, processid: ProcessId) -> CommandReturn {
+        self.apps
+            .enter(processid, |app, _| {
+                // Unconditionally mark this client as subscribed so it will get
+                // a callback when we get the distance reading.
+                app.subscribed = true;
+
+                // If we do not already have an ongoing read, start one now.
+                if !self.busy.get() {
+                    self.busy.set(true);
+                    match self.driver.read_distance() {
+                        Ok(()) => CommandReturn::success(),
+                        Err(e) => CommandReturn::failure(e),
+                    }
+                } else {
+                    // Just return success and we will get the upcall when the
+                    // distance read is ready.
+                    CommandReturn::success()
+                }
+            })
+            .unwrap_or_else(|err| CommandReturn::failure(err.into()))
+    }
+}
+
+impl<'a, T: hil::sensors::Distance<'a>> hil::sensors::DistanceClient for DistanceSensor<'a, T> {
+    fn callback(&self, distance_val: Result<u32, ErrorCode>) {
+        // We completed the operation so we clear the busy flag in case we get
+        // another measurement request.
+        self.busy.set(false);
+
```

Comment:
What if the measurement failed, such as a timeout? Busy will be cleared, but apps won't receive any callback, and their `app.subscribed` will still be set to true.

Reply:
I’ll add an error callback if there’s a timeout or failure so the app knows and `app.subscribed` gets cleared properly. Thanks